### ISSUE TITLE
Include everyone on the ticket, CC & To emails - Enhancment

### ISSUE THREAD

**Issue by ghost** [osTicket/osTicket-1.7#506]
It would be great if when some one sends a support email and includes a colleague or several colleagues in the To or CC fields, the colleague(s) were included in all future notifications relating to the ticket.

Not everyone would want this, so I guess it should be something you can turn on and off in the settings. Maybe also give recipents a link to opt out of further notifications.

Anyway, I'm working on this in my fork https://github.com/WalterEgo/osTicket-1.7-MultipleEmailRecipients because I need it for the folks I'm working for just now and I've just about got a minimum viable product version working.

Not sure how to liaise with whoever oversees the osTicket project as a whole (Peter Rotich) so I can make sure I do it in the best way possible for easy inclusion in osTicket proper?

Will

**Comment by jeffchulg** [comment]
One point that should be done too in the same idea... is that all CC colleagues could login using their email address and the shared ticket id.

**Comment by protich** [comment]
I had a long list of requirements for the feature but I can't find it at the moment. Anyways - below are my quick thoughts, to get the discussion started. 

Implementing CC/BCC code-wise is trivial - but what you need to worry the most about are privacy/sharing related issues because of how osTicket implements authentication, and any implications on the rest of the system.
1. Logging using ticket ID & email to view ALL previous tickets would need to be disabled.
   - username & password  makes sense to view all tickets for the original requestor.
2. You'll need to use different authentication codes - for auto-login
   - System would need to validate the logged in user - think removed CC.
   - Web interface might have different views or options based on who is logged in.
3. You'll need to use a different template to CC - basically the system will do BCC on outgoing emails.
   - You want replies from CCed parties to come back to osTicket - and let osTicket handle alerts out with opt-out options.
   - Template can have a variable for names of CCed parties but not emails.
4. Replies from CCed parties would need to be broadcasted out as alerts
   - Allow CCed parties to opt-out.
5. You'll need to track who was CCed and any corresponding replies in the ticket thread.
   - This can get a little interesting when you factor in emailed replies.

I'm sure I'm forgetting dozens of other issues to consider - but hopefully above will give you an idea on what it would take to implement the feature.

**Comment by ghost** [comment]
Yes, I clocked the viewing all tickets thing last night just before I decided going home and eating trumped further coding. It's the most fundamental thing to sort before my mods are used in a production environment.

I'm whacking things to do as I come accross them on https://github.com/WalterEgo/osTicket-1.7-MultipleEmailRecipients/issues so I don't forget, feel free to add anything else as the come into your head.

**Comment by ghost** [comment]
Quick question: What is the best way to document any additions I make to the database?

**Comment by protich** [comment]
@WalterEgo - We use a monkey-patching like system to allow for continuous DB changes at any point of development cycle. See https://github.com/protich/osTicket-1.7/commit/14efd716b67e83e3cef1da6ddbf17109f0d992dd as an example of a pull request with database changes.
- Schema signature is a md5 hash of core osTicket SQL schema (what new installation use) -  setup/inc/sql/osTicket-mysql.sql 
- Any schema changes must include upgrade path (for existing installations) in a form of patch.
- The upgrade patch name is based on the first 8 chars of FROM-TO signatures.

At first glance it looks complicated but the benefit of not having db changes be bottleneck of development is crazily-awesome. It's one of the features we added in v1.7 that I'm very proud of... although documentation is a little lacking at the moment.

**Comment by ghost** [comment]
Cool, I'll keep my SQL in a temporary file for now and monkey-patch it up when I'm done.

**Comment by ghost** [comment]
Re: but what you need to worry the most about are privacy/sharing related issues because of how osTicket implements authentication, and any implications on the rest of the system

I think I've got this covered now sort of...

I made it omit the email part of the link when emailing when folks are CC'd and you just have to enter you email when you click the link.

I made changes to login, ticket select queries etc. so that it knows you should have access to tickets where you are in the CC emails.

Then I'm having a thought... if you put someone else's email in you will still have access to their tickets... maybe a password should be added too, but then that requires someone to register and stuff. I might let that one mull while I get on with some actual work.

**Comment by ghost** [comment]
Oh, has it really been really two weeks since I last looked at this!

I think I'm going to make it create individual hashes for each person CC'd on the ticket and send them separate emails, each with their own login URL.

Any comments or thoughts before I start stitching this in are most welcome.

**Comment by nate-double-u** [comment]
Noob question...

I've installed osTicket 1.7 and have it running live. I have been wanting this kind of functionality, but am wondering how to install this patch--can this go in over an existing install with existing data in the database?

**Comment by ghost** [comment]
This isn't quite finished yet, there's a security issue I need to deal with, but paying client projects have consumed all my time lately. We were talking about getting this finished at work the other day, so maybe some time will be found in June.

**Comment by nate-double-u** [comment]
Ah, that makes sense, thx for the response. I'll keep a watch here, It'll be a great addition :)
